### PR TITLE
KIBA-1711: Create new tokens count api

### PR DESCRIPTION
**Pull request by Femi-Ogunkola**
Screenshot 2023-03-14 at 12 39 47

ALTER TABLE tbl_collection_hourly_activities
ADD minted_token_count NUMERIC(256,0);

12_HOURS: 0.59s
24_HOURS: 0.55s
7_DAYS: 0.65s
30_DAYS: 0.7s

```
{
  "mintedTokenCounts": [
    {
      "date": "2021-02-11T00:00:00",
      "count": "57529861930983781478"
    },
    {
      "date": "2021-02-11T00:00:00",
      "count": "390583512623193951476867"
    },
    {
      "date": "2021-02-11T00:00:00",
      "count": "6439226427009197034356"
    },
    {
      "date": "2021-02-11T00:00:00",
      "count": "2668808064982186225048915"
    },
    {
      "date": "2021-02-11T00:00:00",
      "count": "9739787688228260894505"
    },
....]
```

### CODE REVIEW

File: api/notd/store/schema.py [hunk 1]
```diff
@@ -139,6 +139,7 @@
     sqlalchemy.Column(key='address', name='address', type_=sqlalchemy.Text, nullable=False),
     sqlalchemy.Column(key='date', name='date', type_=sqlalchemy.DateTime, nullable=False),
     sqlalchemy.Column(key='transferCount', name='transfer_count', type_=sqlalchemy.Numeric(precision=256, scale=0), nullable=False),
+    sqlalchemy.Column(key='mintCount', name='mint_count', type_=sqlalchemy.Numeric(precision=256, scale=0), nullable=False),
```
Comment (krishan711) [inline api/notd/store/schema.py:142]: you need to update the actual schema

Comment (krishan711) [reply api/notd/store/schema.py:142]: this all needs to be optional for now cos its gonna take AGES to back fill this. Make another PR (wip) that sets it back to not optional so I can merge that once the backfilling is done.

Comment (Femi-Ogunkola) [reply api/notd/store/schema.py:142]: Okay

File: api/notd/api/models_v1.py [hunk 2]
```diff
@@ -288,3 +288,7 @@ class ApiSuperCollectionOverlap(BaseModel):
     otherRegistryAddress: str
     otherRegistryTokenCount: int
     registryTokenCountMap: Dict[str, int]
+
+class ApiMintedTokenCount(BaseModel):
+    date: datetime.date
+    mintedTokenCount: Optional[int]
```
Comment (krishan711) [inline api/notd/api/models_v1.py:294]: this one shouldnt be optional - it should just count so be 0 if there are non filled

File: api/notd/model.py [hunk 3]
```diff
@@ -279,6 +279,7 @@ class RetrievedCollectionHourlyActivity:
     date: datetime.datetime
     transferCount: int
     saleCount: int
+    mintCount: int
```
Comment (krishan711) [inline api/notd/model.py:282]: these need to be optional

File: api/notd/model.py [hunk 4]
```diff
@@ -669,3 +671,9 @@ class GallerySuperCollectionUserRow:
     uniqueOwnedTokenCountMap: Dict[str, int]
     chosenOwnedTokensMap: Dict[str, List[TokenMetadata]]
     galleryBadgeHolders: List[GalleryBadgeHolder]
+
+
+@dataclasses.dataclass
+class MintedTokenCount():
+    date: datetime.date
+    mintedTokenCount: Optional[int]
```
Comment (krishan711) [inline api/notd/model.py:679]: this shouldnt be optional

File: api/notd/store/schema.py [hunk 5]
```diff
@@ -139,6 +139,7 @@
     sqlalchemy.Column(key='address', name='address', type_=sqlalchemy.Text, nullable=False),
     sqlalchemy.Column(key='date', name='date', type_=sqlalchemy.DateTime, nullable=False),
     sqlalchemy.Column(key='transferCount', name='transfer_count', type_=sqlalchemy.Numeric(precision=256, scale=0), nullable=False),
+    sqlalchemy.Column(key='mintCount', name='mint_count', type_=sqlalchemy.Numeric(precision=256, scale=0)),
```
Comment (krishan711) [inline api/notd/store/schema.py:142]: optional

File: api/notd/store/schema.py [hunk 6]
```diff
@@ -155,6 +156,7 @@
     sqlalchemy.Column(key='updatedDate', name='updated_date', type_=sqlalchemy.DateTime, nullable=False),
     sqlalchemy.Column(key='address', name='address', type_=sqlalchemy.Text, nullable=False),
     sqlalchemy.Column(key='transferCount', name='transfer_count', type_=sqlalchemy.Numeric(precision=256, scale=0), nullable=False),
+    sqlalchemy.Column(key='mintCount', name='mint_count', type_=sqlalchemy.Numeric(precision=256, scale=0), nullable=False),
```
Comment (krishan711) [inline api/notd/store/schema.py:159]: optional

File: api/notd/store/schema_conversions.py [hunk 7]
```diff
@@ -182,6 +183,7 @@ def collection_total_activity_from_row(rowMapping: RowMapping) -> CollectionTota
         updatedDate=rowMapping[CollectionTotalActivitiesTable.c.updatedDate],
         address=rowMapping[CollectionTotalActivitiesTable.c.address],
         transferCount=int(rowMapping[CollectionTotalActivitiesTable.c.transferCount]),
+        mintCount=int(rowMapping[CollectionTotalActivitiesTable.c.mintCount]),
```
Comment (krishan711) [inline api/notd/store/schema_conversions.py:186]: both of these should be optional

File: api/notd/manager.py [hunk 8]
```diff
@@ -575,3 +577,30 @@ async def update_token_staking_deferred(self, registryAddress: str, tokenId: str
...
+        mintedTokenCount = sqlalchemyfunc.sum(CollectionHourlyActivitiesTable.c.mintCount).label('mintedTokenCount') # type: ignore[no-untyped-call, var-annotated]
+        query = (
+            sqlalchemy.select(date, mintedTokenCount)
+            .where(CollectionHourlyActivitiesTable.c.date >= endDate)
+            .where(CollectionHourlyActivitiesTable.c.date < currentDate)
+            .group_by(date)
+            .order_by(date.desc())
+        )
+        result = await self.retriever.database.execute(query=query)
+        mintedTokenCounts = [MintedTokenCount(date=day['date'], mintedTokenCount=day['mintedTokenCount']) for day in result.mappings()]
```
Comment (krishan711) [inline api/notd/manager.py:605]: will this return a value for EVERY day in the duration, including days with no results?

File: api/notd/manager.py [hunk 9]
```diff
@@ -575,3 +577,30 @@ async def update_token_staking_deferred(self, registryAddress: str, tokenId: str
...
+            endDate = date_util.datetime_from_datetime(dt=currentDate, days=-30)
+        elif duration == '24_HOURS':
+            date = CollectionHourlyActivitiesTable.c.date.label('date')
+            endDate = date_util.datetime_from_datetime(dt=currentDate, hours=-24)
+        elif duration == '12_HOURS':
+            date = CollectionHourlyActivitiesTable.c.date.label('date')
+            endDate = date_util.datetime_from_datetime(dt=currentDate, hours=-12)
+        else:
+            raise BadRequestException('Unknown duration')
+        mintedTokenCount = sqlalchemyfunc.sum(CollectionHourlyActivitiesTable.c.mintCount).label('mintedTokenCount') # type: ignore[no-untyped-call, var-annotated]
```
Comment (krishan711) [inline api/notd/manager.py:596]: you can inline this i think?

File: api/notd/api/models_v1.py [hunk 10]
```diff
@@ -288,3 +288,7 @@ class ApiSuperCollectionOverlap(BaseModel):
     otherRegistryAddress: str
     otherRegistryTokenCount: int
     registryTokenCountMap: Dict[str, int]
+
+class ApiMintedTokenCount(BaseModel):
+    date: datetime.date
+    mintedTokenCount: int
```
Comment (krishan711) [inline api/notd/api/models_v1.py:294]: same comment here as below in model.py

File: api/notd/api/models_v1.py [hunk 11]
```diff
@@ -288,3 +288,7 @@ class ApiSuperCollectionOverlap(BaseModel):
     otherRegistryAddress: str
     otherRegistryTokenCount: int
     registryTokenCountMap: Dict[str, int]
+
+class ApiMintedTokenCount(BaseModel):
+    date: Union[datetime.datetime, datetime.date]
```
Comment (krishan711) [inline api/notd/api/models_v1.py:293]: this should only ever be a datetime.datetime

Comment (krishan711) [reply api/notd/api/models_v1.py:293]: (should be obiovous cos we NEVER have Union[date, datetime] anywhere in a model

File: api/notd/api/response_builder.py [hunk 12]
```diff
@@ -500,3 +502,12 @@ async def collection_overlap_owner_from_model(self, collectionOverlapOwner: Coll
 
     async def collection_overlap_owners_from_models(self, collectionOverlapOwners: Sequence[CollectionOverlapOwner]) -> List[ApiCollectionOverlapOwner]:
         return await asyncio.gather(*[self.collection_overlap_owner_from_model(collectionOverlapOwner=collectionOverlapOwner) for collectionOverlapOwner in collectionOverlapOwners])
+
+    async def minted_token_count_from_model(self, mintedTokenCount: MintedTokenCount) -> ApiMintedTokenCount:
+        return ApiMintedTokenCount(
+            date=mintedTokenCount.date,
+            count=str(mintedTokenCount.count) if mintedTokenCount.count else "0"
```
Comment (krishan711) [inline api/notd/api/response_builder.py:509]: why is this a sstring?! and in what cases would this if condition be executed?

**Comment by krishan711** [comment]
@Femi-Ogunkola your api response above is definitely wrong. why are all the counts 0?? you need to run this on real data in order to get real times and real outputs.

File: api/notd/api/models_v1.py (hunk 11 above)
Comment (Femi-Ogunkola) [reply api/notd/api/models_v1.py:293]: I did it like this cause depending on the duration, we can either have days ie "2022-12-31" or datetime "2021-02-11T13:00:00" in the response. Or for days should it always be "2022-12-31T00:00:00"

Comment (krishan711) [reply api/notd/api/models_v1.py:293]: yes we store dates as datetimes everywhere so should just return this outwards. optional types are super confusing for whoever is consuming the api

**Pull request merged**